sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala

```diff
@@ -170,7 +170,7 @@ object FileFormatWriter extends Logging {
             description = description,
             sparkStageId = taskContext.stageId(),
             sparkPartitionId = taskContext.partitionId(),
-            sparkAttemptNumber = taskContext.attemptNumber(),
+            sparkAttemptNumber = taskContext.taskAttemptId().toInt & Integer.MAX_VALUE,
```

Comment:
After more than 2 years, I revisited this code path again and realized that this is not the best fix.

The original motivation is still correct: Spark violates the contract of `TaskAttemptID`, as Spark resets task attempt number after stage retry, making `TaskAttemptID` not unique.

The root cause is: Spark job has stages while Hadoop job directly has tasks (no DAG). We map Spark stage id to Hadoop job id, which is inaccurate as this doesn't count stage attempt number.

I think a better fix is to generate the Hadoop job id with both stage id and stage attempt number, or generate the Hadoop task attempt number with both Spark task and stage attempt number. However, the current fix also works, as this only decides the intermedia staging directory name, which we don't care as long as it's unique.

I'm leaving this comment just for future references.